Content.Client/BarSign/Ui/BarSignMenu.xaml.cs

```diff
@@ -0,0 +1,64 @@
+﻿using System.Linq;
+using Content.Client.UserInterface.Controls;
+using Content.Shared.BarSign;
+using Robust.Client.AutoGenerated;
+using Robust.Client.UserInterface.XAML;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Timing;
+
+namespace Content.Client.BarSign.Ui;
+
+[GenerateTypedNameReferences]
+public sealed partial class BarSignMenu : FancyWindow
+{
+    [Dependency] private readonly IEntityManager _entity = default!;
+    [Dependency] private readonly IPrototypeManager _prototype = default!;
+
+    private readonly EntityUid _owner;
+    private string? _currentId;
+
+    private readonly List<BarSignPrototype> _cachedPrototypes = new();
+
+    public event Action<string>? OnSignSelected;
+
+    public BarSignMenu(EntityUid owner)
+    {
+        RobustXamlLoader.Load(this);
+        IoCManager.InjectDependencies(this);
+
+        _owner = owner;
+        _currentId = _entity.GetComponentOrNull<BarSignComponent>(owner)?.Current;
+
+        _cachedPrototypes.Clear();
+        _cachedPrototypes = Shared.BarSign.BarSignSystem.GetAllBarSigns(_prototype)
+            .OrderBy(p => Loc.GetString(p.Name))
+            .ToList();
+        foreach (var proto in _cachedPrototypes)
+        {
+            SignOptions.AddItem(Loc.GetString(proto.Name));
+        }
+
+        SignOptions.OnItemSelected += idx =>
+        {
+            OnSignSelected?.Invoke(_cachedPrototypes[idx.Id].ID);
+            SignOptions.SelectId(idx.Id);
+        };
+
+        if (_currentId != null)
+        {
+            var idx = _cachedPrototypes.IndexOf(_prototype.Index<BarSignPrototype>(_currentId));
+            SignOptions.TrySelectId(idx);
+        }
+    }
+
+    protected override void FrameUpdate(FrameEventArgs args)
+    {
+        base.FrameUpdate(args);
+
+        if (_entity.GetComponentOrNull<BarSignComponent>(_owner)?.Current is not { } id || id.Id == _currentId)
+            return;
+        _currentId = id.Id;
+        var idx = _cachedPrototypes.IndexOf(_prototype.Index(id));
+        SignOptions.TrySelectId(idx);
+    }
```

Comment:
This looks like a hack, what are you trying to fix here? If its to make the sign change when the server dirtys the component that's pretty bad, just use the normal UpdateState pattern